feat(groups): Implement domain-based group assignment and management

Locutus7 opened the pull request:
# Jira https://dsai.atlassian.net/browse/CHAT-897

Locutus7 commented on src/lib/components/admin/Users/Groups/Users.svelte:
```diff
@@ -46,7 +47,32 @@
+				</svg>
+				<div class="text-sm text-blue-700 dark:text-blue-300">
+					{$i18n.t(
+						'Individual user selection is disabled when domains are selected for this group.'
```
It's primarily about avoiding conflicting access logic and maintaining clear administrative intent. When domains are selected, the system automatically grants access to all users from those domains (e.g., everyone with @agr.gc.ca gets access). If we also allowed individual user selection, it creates ambiguity: does a user from @cra-arc.gc.ca get access because they're individually selected, or are they excluded because their domain isn't in the allowed list? This could lead to unexpected access patterns where admins think they've restricted access to certain domains but individual selections override that intent.

The current design enforces clear access patterns: either you're doing department-wide access (via domains) or individual user curation. This makes it easier for admins to understand and audit who has access and why. However, this could potentially be enhanced to support additive access (domains + individual users) with clear precedence rules, where domain access is the base level and individual selections add additional users from non-included domains. The current approach prioritizes simplicity and predictability over flexibility.

Locutus7 replied:
Fixed.

justbert commented on backend/open_webui/migrations/versions/64d8cd268d63_add_allowed_domains_to_groups_table.py:
```diff
@@ -0,0 +1,29 @@
+
+# revision identifiers, used by Alembic.
+revision: str = "64d8cd268d63"
+down_revision: Union[str, None] = "a71ba3c6d0a4"
```
We're gonna need to be careful when merging here due to other PRs:
https://github.com/ssc-dsai/canchat-v2/pull/232/files#diff-913fbda72496fd26dc46acee231a845f05e62f92f7e883a3285c11a4a066264bR18

justbert reviewed (commented):
It might be nice to have a search when adding domains to groups?

justbert commented on backend/open_webui/models/groups.py:
```diff
@@ -2,18 +2,50 @@
+
+def _invalidate_group_membership_cache():
+    """Invalidate the entire group membership cache."""
+    global _group_membership_cache, _cache_timestamp
```
Awesome! Appreciate the info, still getting my Python legs in place!

justbert commented:
Just wondering if you wanted to respond to these:
- [Should be move Domain management closer to Users and Groups?](https://github.com/ssc-dsai/canchat-v2/pull/228#issuecomment-3333526222) 
- https://github.com/ssc-dsai/canchat-v2/pull/228#discussion_r2372324522
- https://github.com/ssc-dsai/canchat-v2/pull/228#discussion_r2372300107

Locutus7 commented on src/lib/components/admin/Settings/Domains.svelte:
```diff
@@ -0,0 +1,394 @@
+		}
+
+		// Domain validation - Government of Canada domains (*.gc.ca + special cases)
+		const domainRegex =
+			/^([a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.gc\.ca|elections\.ca|canada\.ca|gc\.ca)$/;
+		if (!domainRegex.test(domainName)) {
+			toast.error(
+				i18n.t(
+					'Please enter a valid Government of Canada domain (*.gc.ca, canada.ca, elections.ca)'
+				)
+			);
+			return;
+		}
```
Fixed.

Locutus7 commented:
> > > > > > > A couple of questions so I can get some of the design decisions in my brain :) :
> > > > > > > 
> > > > > > > * Is there a way to add domains via the UI?
> > > > > > > * Why did you choose to create a table for domains, specifically? We aren't using their IDs in the group access, so it seems like a large amount of extra code for lists that are added directly to the `Group` model
> > > > > > 
> > > > > > 
> > > > > > Q: Is there a way to add domains via the UI?
> > > > > > Yes, there is a comprehensive UI for domain management. The system includes:
> > > > > > Admin Settings → Domains Tab: Full CRUD interface located at Domains.svelte
> > > > > > 
> > > > > > * Add new domains with validation
> > > > > > * Edit existing domain names and descriptions
> > > > > > * Delete domains with confirmation
> > > > > > * View all configured domains in a list
> > > > > > 
> > > > > > Group Edit Modal → Domain Selection: Simplified selection interface at Domains.svelte
> > > > > > 
> > > > > > * Select from available domains for group assignment
> > > > > > * Links to admin settings when no domains are configured
> > > > > > 
> > > > > > Q: Why create a table for domains instead of using lists directly in the Group model?
> > > > > > Several architectural and functional reasons justify the separate domains table:
> > > > > > Data Integrity & Validation
> > > > > > 
> > > > > > * Centralized validation: All domains are validated once when added, ensuring consistent Government of Canada domain format
> > > > > > * Referential integrity: Prevents typos and invalid domains in group assignments
> > > > > > * Data normalization: Eliminates duplicate domain storage across multiple groups
> > > > > > 
> > > > > > Administrative Control
> > > > > > 
> > > > > > * Admin-managed whitelist: Only administrators can add/remove approved domains through the dedicated UI
> > > > > > * Audit trail: Each domain has creation/update timestamps and descriptions for tracking
> > > > > > * Bulk management: Easy to see all configured domains, edit descriptions, and manage the approved list
> > > > > > 
> > > > > > User Experience
> > > > > > 
> > > > > > * Domain selection UI: Groups can select from a dropdown of valid domains rather than typing free-form text
> > > > > > * Descriptive labels: Domains include descriptions (e.g., "Agriculture and Agri-Food Canada" for "agr.gc.ca")
> > > > > > * Error prevention: Users cannot accidentally assign invalid or unapproved domains
> > > > > > 
> > > > > > Future Extensibility
> > > > > > 
> > > > > > * Additional metadata: Can easily add fields like contact info, approval status, or expiration dates
> > > > > > * Domain-specific permissions: Could potentially add domain-level access controls
> > > > > > * Reporting capabilities: Can analyze domain usage across groups
> > > > > 
> > > > > 
> > > > > The reason I was asking about how to add Domains is because I can't see this tab anywhere: <img alt="image" width="1779" height="468" src="https://private-user-images.githubusercontent.com/6690907/492845416-ea4d657c-2aa7-4a6d-8b51-ff89c1ae2f8e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg2NDM2NzQsIm5iZiI6MTc1ODY0MzM3NCwicGF0aCI6Ii82NjkwOTA3LzQ5Mjg0NTQxNi1lYTRkNjU3Yy0yYWE3LTRhNmQtOGI1MS1mZjg5YzFhZTJmOGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MjNUMTYwMjU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTcyOTlmOWIzYTlkODRlNzA0ZTU1MzU1OWZiM2I3ZTdhZDA0ZjBkZDhlYmIwMmJlNWIxZjQ2ZTgwOTg4ZmM3NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.1Kgpo_PLEwtMiSAfB8-nH61tQYlJ9hj4mtb4-QYyA1Q">
> > > > > I assumed it would be under the Users and Groups sections. Was there a reason you chose Settings?
> > > > > I was just wondering about the complexity since we will likely not be continuing the development too far in the future due to other priorities but I appreciate the response.
> > > > 
> > > > 
> > > > Look in Groups then either choose an existing one or create one. Everything is in there.
> > > 
> > > 
> > > I don't follow. That doesn't answer my question.
> > 
> > 
> > Let me clarify the UI locations.
> > Domain Management is in TWO places:
> > 
> > 1. Admin Settings → Domains Tab (for system-wide domain management)
> > 
> > 
> > * Navigate to Admin Settings (gear icon in sidebar)
> > * Look for the "Domains" tab alongside General, Users, etc.
> > * This is where administrators add/edit/delete the master list of approved domains
> > 
> > 
> > 2. Groups → Edit Group → Domains Tab (for selecting domains for a specific group)
> > 
> > 
> > * Go to Admin → Users → Groups
> > * Edit an existing group or create a new one
> > * In the group edit modal, there's a "Domains" tab where you select which domains this group allows
> > 
> > The reason for putting domain management in Admin Settings rather than the Groups section:
> > 
> > * Domain management is a system-wide administrative function
> > * Similar to how user management is in Admin Settings, not embedded within individual features
> > * Keeps the approved domain list centralized and controlled by administrators
> > * Groups section focuses on group-specific configuration (selecting from approved domains)
> > 
> > If you're not seeing the Domains tab in Admin Settings, it might be a permissions issue or the tab hasn't been added to your build yet. Regarding complexity - I understand the future priorities concern. The domain system can be simplified if needed by removing the separate table and using simple string arrays in groups instead.
> 
> Thanks for the overview! I think my questions are coming more from the placement of the Domains tab. I can see the tab, my original search was to look for it with the users config. I think I'm trying to understand why it was in the `Settings` tab which deals with system and functionality settings instead of the `Users` tab which deals with IAM such as user and groups administration. It feels like the `Users` tab is more of a similar underlying concept to manage domains in. We can always change the name of the `Users` tab to something more descriptive; it hasn't changed and with the addition of `Groups`, it's not just user management anymore but management related to users? Being a lazy administrator, that's a lot of extra clicks if I want to add a new domain to set in a group!
> 
> As for reverting to another solution, I don't mind this at all, I just hadn't been able to find the place to add domains and was a bit confused.

Fixed.

Locutus7 commented:
> Quote reply

Thanks for the comprehensive review, as always!
